src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Routing.Internal;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class RoutePatternMatcher
+    {
+        private const string SeparatorString = "/";
+        private const char SeparatorChar = '/';
+
+        // Perf: This is a cache to avoid looking things up in 'Defaults' each request.
+        private readonly bool[] _hasDefaultValue;
+        private readonly object[] _defaultValues;
+
+        private static readonly char[] Delimiters = new char[] { SeparatorChar };
+
+        public RoutePatternMatcher(
+            RoutePattern pattern,
+            RouteValueDictionary defaults)
+        {
+            if (pattern == null)
+            {
+                throw new ArgumentNullException(nameof(pattern));
+            }
+
+            RoutePattern = pattern;
+            Defaults = defaults ?? new RouteValueDictionary();
```

Comment:
Allowing a null defaults feels weird.

Maybe a ctor that takes just pattern and another that takes both and checks defaults to not be null?

Reply:
See your comment else where about `Exception(string message)`. It's normal to accept null for a parameter that's optional. 

I think the methods in MVC for `MapRoute` also will tolerate null, so it would require extra work all over the code 😆